Description: add smart mirror patch
Author:  electricface <songwentai@deepin.org>

---
The information above should follow the Patch Tagging Guidelines, please
checkout http://dep.debian.net/deps/dep3/ to learn about the format. Here
are templates for supplementary fields that you might want to add:

Origin: <vendor|upstream|other>, <url of original patch>
Bug: <url in upstream bugtracker>
Bug-Debian: https://bugs.debian.org/<bugnumber>
Bug-Ubuntu: https://launchpad.net/bugs/<bugnumber>
Forwarded: <no|not-needed|url proving that it has been forwarded>
Reviewed-By: <name and email of someone who approved the patch>
Last-Update: 2019-10-30

Index: v23/apt-pkg/acquire-worker.cc
===================================================================
--- v23.orig/apt-pkg/acquire-worker.cc	2022-11-04 10:15:25.089268437 +0800
+++ v23/apt-pkg/acquire-worker.cc	2022-11-04 10:15:25.089268437 +0800
@@ -38,6 +38,7 @@
 
 #include <apti18n.h>
 									/*}}}*/
+#include "smartmirrors.h"
 
 using namespace std;
 
@@ -812,6 +813,7 @@
 
    Item->SyncDestinationFiles();
 
+   Item->URI = SmartMirrors::GuestURI(Item->URI);
    string Message = "600 URI Acquire\n";
    Message.reserve(300);
    URI URL(Item->URI);
Index: v23/apt-pkg/smartmirrors.cc
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ v23/apt-pkg/smartmirrors.cc	2022-11-04 10:16:40.653265339 +0800
@@ -0,0 +1,76 @@
+// -*- mode: cpp; mode: fold -*-
+// Description								/*{{{*/
+/* ######################################################################
+
+   Smart Mirrors - Get Archive File by Smart Mirrors Server.
+
+   ##################################################################### */
+/*}}}*/
+// Include Files							/*{{{*/
+
+
+#include <apt-pkg/acquire-item.h>
+#include <apt-pkg/configuration.h>
+#include <apt-pkg/aptconfiguration.h>
+#include <apt-pkg/fileutl.h>
+
+#include <tr1/memory>
+#include "smartmirrors.h"
+
+using namespace std;
+
+namespace SmartMirrors {
+        static string
+        normalizeURI(const string& uri)
+        {
+                if (uri.length() >=1 && uri.at(uri.length()-1) == '/') {
+                        return uri.substr(0, uri.length()-1);
+                }
+                return uri;
+        }
+
+        const std::string
+        GuestURI(const std::string& uri)
+        {
+                if(uri.substr(0, strlen("http")) != "http") {
+                        return uri;
+                }
+                static bool debug = _config->FindB("Acquire::SmartMirrors::Debug");
+                static bool enabled =  _config->FindB("Acquire::SmartMirrors::Enable");
+                if (!enabled) {
+                        if (debug) {
+                                std::clog << "SmartMirrors is disabled." << std::endl;
+                        }
+                        return uri;
+                }
+                static string official = normalizeURI(_config->Find("Acquire::SmartMirrors::MainSource"));
+                static string mirror = normalizeURI(_config->Find("Acquire::SmartMirrors::MirrorSource"));
+
+                static string detector = _config->Find("Acquire::SmartMirrors::GuestURI");
+                static bool exists = FileExists(detector);
+                if (!exists) {
+                        return uri;
+                }
+
+                string cmd = detector + " " + uri + " " + official + " " + mirror;
+                std::tr1::shared_ptr<FILE> out(popen(cmd.c_str(), "r"), pclose);
+                if (!out) {
+                        return uri;
+                }
+
+                char buffer[1024] = {0};
+                std::string result = "";
+                while (!feof(out.get())) {
+                        if (fgets(buffer, 1024, out.get()) != NULL)
+                                result += buffer;
+                }
+                if(result.substr(0, strlen("http")) != "http") {
+                        return uri;
+                }
+
+                if (debug && result != uri) {
+                        printf("Using '%s' instead of '%s'\n", result.c_str(), uri.c_str());
+                }
+                return result;
+        }
+}
Index: v23/apt-pkg/smartmirrors.h
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ v23/apt-pkg/smartmirrors.h	2022-11-04 10:15:25.089268437 +0800
@@ -0,0 +1,8 @@
+#include <config.h>
+
+#include <apt-pkg/cacheiterators.h>
+#include <apt-pkg/indexfile.h>
+
+namespace SmartMirrors {
+  const std::string GuestURI(const std::string& uri);
+};
